### PR TITLE
[draft] tree test code

### PR DESCRIPTION
**Pull request by mprast**
# XRX Tree Tester
This PR adds a library that for instrumenting and evaluating the performance (re: correctness, not speed) of xRx in certain scenarios. There is a library called TreeTestRunner that can be used to a) create a hierarchy of nodes representing the agent execution flow b) run an LLM to answer a series of qualitative questions about each node to determine if the agent successfully completed the scenario. The idea is that by representing agent execution as a tree, we can get fine-grained information about the success or failure of each node, and whether and how those nodes contributed to the (in)ability of the agent to successfully complete the scenario.

For demonstration purposes, I've use the TreeTestRunner to instrument `shopify-app`. TreeTestRunner supports arbitrary trees; the structure I've chosen is:

    - one root node at the top level, representing a conversation between the user and the agent
    - many request-response nodes at the second level. Each of these nodes captures a message from the user to the agent and an associated response from the agent back to the user.
    - many graph-specific nodes below each request-response node. Each of these corresponds to a specific reasoning task the agent undertakes while formulating a response to the user (`routing`, `chooseTool`, etc.)

# Testing
To see TreeTestRunner in action, start shopify-app and interact with the app normally (I'd recommend using `interactive_test.py`.) When you are finished, say or type 'stop'. Wait for a bit for the tests to run (this may take a while, especially if you get rate-limited by the LLM provider.) After the tests are run the frontend will crash, but you should see 

```
xrx-reasoning     | 2024-11-09 21:53:19,146 INFO:Done writing file!
xrx-reasoning     | 2024-11-09 21:53:19,149 INFO:Done writing simple file!
```

in the docker-compose output. At this point you can find the output of your test in `shopify-app/testOutput`. There will be two files, one of the form `<name>.json` and one of the form `<name>-simple.json`. The first one is raw output and the second is cleaned up a bit to be more pleasant to read. I recommend using a json prettifier to look at these.

The description of the tests to run is in `shopify-app/reasoning/app/treeTestConfig/buySomethingTest.yaml`. It includes the goal that the agent is supposed to fulfill, as well as a list of audits for every possible stage. The format should be self-explanatory, but just leave a comment if it's not and I can document it. Feel free to add config for your own scenarios; you can change the test being run by changing the pathname in line 40 of `executor.py`

You can find some sample outputs under `shopify-app/reasoning/app/agent/sample_tests`. These were all generated by me interacting with the agent.

# TODOs & Cleanup Work
- fix naming - I'm used to typescript these days so I keep reflexively using camel case instead of snake case
- Perhaps find a way to persist the test runner across requests that doesn't involve pickling
- remove various debug cruft (pdb imports)

# Next Steps
- ~Build an LLM-backed "synthetic user" that can interact with the agent based on a prompt~
- Push the tests to S3 instead of writing them locally
- Come up with a good way to summarize multiple tests - the evaluations are predictably stochastic so we'll likely need to run them lots of times to get a stable signal.
- Modify this to accommodate "black box" approaches where we aren't able to instrument the source and instead have to parse the raw output of the system into a tree.
- Anything else I'm forgetting!

### CODE REVIEW

**Comment by alessandro-neri** [comment]
**[Feature] Add Comprehensive Testing Utilities to xRx Framework**

## 1. Overview
- **What is the feature?**  
  This feature introduces a new `testing.py` module to the xRx framework, providing robust utilities for building and managing tests for AI-powered reasoning systems. Additionally, the `__init__.py` file is updated to integrate these testing utilities into the framework.

- **What changed?**
  - Framework Initialization: Enhanced the `__init__.py` to include the new testing utilities.
  - Testing Utilities: Added a comprehensive `testing.py` module with classes and functions for defining and executing tests.

## 2. Files Modified
| File Name | Changes |
|-----------|---------|
| `__init__.py` | • **Context**: Initializes the xRx agent framework.<br>• **Changes [EDIT]**: Added an import statement for the new `testing.py` module to integrate testing functionalities into the framework. |
| `testing.py` | • **Context**: Provides testing utilities for the xRx framework.<br>• **Changes [NEW]**: Introduced `testing.py` with 274 lines of code, including classes like `TestResult`, `TaskNode`, and `TreeTestRunner`, as well as functions for configuration management, prompt generation, and audit execution using LLMs. |

## 3. Issues/Improvements

<details>
<summary>Security. Ensure secure handling of test configurations.</summary>
- Sensitive data in test configurations might be exposed.
- Implement encryption or secure access controls for configuration files.
</details>

<details>
<summary>Performance. Assess the impact of testing utilities on framework performance.</summary>
- Extensive testing operations could slow down the framework.
- Optimize testing processes and consider asynchronous execution where possible.
</details>

<details>
<summary>Maintainability. Manage increased complexity due to added testing modules.</summary>
- The new `testing.py` introduces substantial code that may complicate maintenance.
- Refactor the testing utilities for better modularity and add comprehensive documentation.
</details>

<details>
<summary>Simplification. Potential to streamline configuration merging logic.</summary>
- The `deep_merge` function may be overly complex for current needs.
- Explore using standard library functions or simplifying the merging logic to enhance readability and efficiency.
</details>

---

**Comment by alessandro-neri** [comment]
Hi Matt, adding some AI generated descriptions to gather feedback and test prompts!

**Comment by mprast** [comment]
> Hi Matt, adding some AI generated descriptions to gather feedback and test prompts!

awesome, ty!

**Comment by sabbyanandan** [comment]
@mprast: This is fantastic work! Honestly, though, I had to seek LLM help to understand the `TreeTestRunner` with an example [I chose [defaultAudits.yaml](https://github.com/8090-inc/xrx-sample-apps/pull/30/files#diff-1dbe8281e6b69950a975c9a5fee1779bf414fb9c1935ecbf87442abc5849305e)], but here's how I finally interpret how the evaluation actually works.

```
So:
Root questions (4 questions): Evaluated once for entire conversation

Stage questions (for each interaction):
requestResponse (1 question)
routing (2 questions)
chooseTool (2 questions)
identifyToolParams (1 question)
customerResponse (2 questions)

Each interaction gets evaluated against 8 questions (the sum of all stage questions), and then the entire conversation gets evaluated against the 4 root questions at the end.
```

Aside from the nits (code/function-level docs), I don't have much commentary other than explaining the functionality and the developer guide for developers to take advantage of this test infrastructure as they build use cases with xRx. I would use  [defaultAudits.yaml](https://github.com/8090-inc/xrx-sample-apps/pull/30/files#diff-1dbe8281e6b69950a975c9a5fee1779bf414fb9c1935ecbf87442abc5849305e) as an example to describe how it works in the guide. That can be a separate and outside of this PR.

File: xrx_agent_framework/xrx_agent_framework/utils/testing.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            # Otherwise, value from dict2 overwrites value from dict1
+            else:
+                merged[key] = value
+        else:
+            # Key not in dict1, just add it
+            merged[key] = value
+
+    return merged
+
+def readYaml(file_path):
```
Comment (sabbyanandan) [inline xrx_agent_framework/xrx_agent_framework/utils/testing.py:60]: Perhaps the file manipulation functions can be grouped or gathered together with additional docs. Same gathering and docs might also be helpful for the LLM util-like functions below.

**Review by sabbyanandan — APPROVED** [review]
Some nits, but otherwise, let's go! :shipit:

**Comment by mprast** [comment]
@sabbyanandan thanks for taking a look! will address comments

sorry for not including more detail on how the test runner works, the general gist is:

- you use the tree runner to build a tree of nodes (nodes are also referred to as _tasks_); the structure of the tree is arbitrary
- each task can succeed or fail
- each task has
    - a _stage_ (and optionally a _substage_) - this identifies the type of the task. The stage/substage is used to look up audit questions in the config file. These can be arbitrary
    - _task data_ - this is an arbitrary object which is used to provide context to the LLM when it's asked to answer audit questions about the task. This is the _only_ information the LLM will have about the task when answering the questions
    - a _compositive_ flag - this indicates whether the task is compositive (see below). By default, tasks are not compositive

- the success or failure of a task is calculated like this:
  - a compositive task simply succeeds if all of its child tasks succeed, and fails otherwise.  The LLM is not consulted to determine whether the task has succeeded or failed. Task data is ignored for compositive tasks.
  - to evaluate the success of a non-compositive task, the LLM is asked a series of audit questions about the task data attached to the task. Audit questions are fetched from configuration based on the stage & substage of the task. If all of the audit questions are answered in the affirmative, the task succeeds, otherwise it fails. A non-compositive task may succeed even if some of its children have failed.

I'll compile this information and add it to the docs. Thanks!

**Comment by mprast** [comment]
you can now turn the tree test runner on or off using an env var. xRx will only use a state machine if TREE_TEST_RUNNER_ON is set to "true"

**Pull request merged**